### PR TITLE
HSEARCH-2786 JSR-352: Split job parameter fetchSize into 2 parameters

### PR DESCRIPTION
**Pull request by mincong-h**
Hi @yrodiere , here's the PR for ticket <https://hibernate.atlassian.net/browse/HSEARCH-2786>.

I changed my initial mind about the splitting: actually we don't need to create 2 parameters explicitly for entity fetching, only `idFetchSize` would be enough:

- The "ID Fetching" during the phrase of "Partition Mapping" can be achieved by using the parameter `idFetchSize`;
- The "Entity Fetching" during the phrase of "Entity Reading" can be achieve by using the `rowsPerPartitions`. Thus, there's no need to create the 2nd parameter, called `entityFetchSize`.

Additionally, I found that the fetch size can actually be negative. So I disabled the positive-integer validation in `JobContextSetupListener`: we only require that the fetch size to be an integer, but any value is acceptable. You might ask: why this parameter must be validated meanwhile it is optional? Because we provide a default value when user does not fill it: so in any case, it should be filled at the JSR-352 batch runtime point of view.

### CODE REVIEW

**Comment by yrodiere** [comment]
I think I disagree with simply removing the `fetchSize` parameter like you did.

First, because if we were to define the fetch size automatically from another value, I think it should rather be `checkpointInterval`. It's more or less what we do in the old mass indexer, as mentioned [there](https://github.com/yrodiere/hibernate-search/pull/21#issuecomment-309377142).
The reason is `checkpointInterval` seems to fit more closely the memory-consumption limitation goal of `fetchSize`:
 * partitioning is only about how parallel you want processing to be, so you might want `rowsPerPartition` to be huge to only have one or two partitions per entity type
 * chunking (checkpoints) is about how memory-intensive you want processing to be (how many documents you keep in memory until you write them to the index), so you definitely don't want to only have one checkpoint per entity type if there are millions of entities.
Keep in mind that you can have multiple checkpoints in a single partition.

Second, because `rowsPerPartition` could legitimately be huge, and because `rowsPerPartition` happens to be (more or less) the number of results we expect from the per-partition query. So setting the fetch size to that value effectively tells Hibernate ORM to fetch all the query results immediately.

Third, because as mentioned [there](https://github.com/yrodiere/hibernate-search/pull/21#issuecomment-309377142), there could be legitimate reasons to want the `checkpointInterval` to be different from the `fetchSize`, most notably you could want the `fetchSize` to be higher than the `checkpointInterval` to mitigate latency in database accesses. And you even gave an example in the javadoc of using a magic value for the `fetchSize` when using MySQL.

To sum up: I don't think it's an improvement over what we had before, but I'd settle for using the `checkpointInterval` as the `fetchSize` if you want. That way we'll be at least as good as the old mass indexer.

Oh, and apart from that, the commits look good :)

**Comment by mincong-h** [comment]
I was searching shortcut for this implementation... but yes, you're right. Partition and chunk are conceptually different, and the DB parameter `entityFetchSize` is much closer to the chunk parameter `checkpointInterval`, which are both memory-related parameters. Meanwhile, they're not the same because there're good ideas to have two distinct variables as you've mentioned [in the comment](https://github.com/yrodiere/hibernate-search/pull/21#issuecomment-309377142).

You proposed 3 names in the JIRA ticket [HSEARCH-2786](https://hibernate.atlassian.net/browse/HSEARCH-2786), and I think `entityFetchSize` is the best. Basically because I prefer shorter name for public API, which might be less verbose and easier to memory. We can still rename it during the final PR.

So I'll split the `fetchSize` into 2 parameters as the initial plan:

- `idFetchSize`
- `entityFetchSize`

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/MassIndexingJob.java [hunk 1]
```diff
@@ -33,7 +33,7 @@
  * 		MassIndexingJob.NAME,
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/MassIndexingJob.java]: This Javadoc example is obsolete. Since it's not related to this ticket, I created another ticket for updating it <https://hibernate.atlassian.net/browse/HSEARCH-2824>

**Comment by yrodiere** [comment]
Looks good! Merging, thanks
